rockcraft/models/project.py

```diff
@@ -64,6 +64,14 @@ class Config:  # pylint: disable=too-few-public-methods
             "_", "-"
         )
 
+    @pydantic.validator("build_on", pre=True)
+    @classmethod
+    def _helpful_error_for_non_list_build_on(cls, val: list[str]|Any) -> list[str]:
+        """Provide helpful error when 'build_on' is not a list."""
+        if not isinstance(val, list):
```

Comment:
Thanks for taking a look!

> Note that val could also validly be None here

This validation hook only gets called if `builds-on:` is actually specified in `rockctaft.yaml` so this will **not** cause any problems whatsoever if `builds-on:` is omitted entirely.

The [official documentation only specifies `list[str]`](https://documentation.ubuntu.com/rockcraft/en/latest/reference/rockcraft.yaml/#platforms-entry-build-on), so from my PoV, both a bare `build-on: <nothing>`, as well as explicit `build-on: null` are clear spec violations which warrant an explicit error.

> regardless, I think the right place to fix this issue is [where handle the platform error](https://github.com/canonical/rockcraft/blob/main/rockcraft/models/project.py#L186)

The [Platform class gets instantiated right above that line](https://github.com/canonical/rockcraft/blob/main/rockcraft/models/project.py#L184), at which point its validators get run, so **by that stage the erroneous error message is already** there and the damage already done.

Reply:
> This validation hook only gets called if builds-on: is actually specified in rockctaft.yaml so this will not cause any problems whatsoever if builds-on: is omitted entirely.

That might be but there are many tests failing with this new check; e.g. `test_image_service_cache`

> The [Platform class gets instantiated right above that line](https://github.com/canonical/rockcraft/blob/main/rockcraft/models/project.py#L184), at which point its validators get run, so by that stage the erroneous error message is already there and the damage already done.

The `error_dict` object has the location of the error, so it should be possible to provide a better message with the info that's already there.